tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs

```diff
@@ -0,0 +1,147 @@
+                {
+                    if (expectedMechanismProperties.IsBsonNull)
+                    {
+                        if (authenticator is GssapiAuthenticator)
+                        {
+                            var serviceName = authenticator._mechanism()._serviceName();
+                            serviceName.Should().Be("mongodb"); // The default is "mongodb".
+                            var canonicalizeHostName = authenticator._mechanism()._canonicalizeHostName();
+                            canonicalizeHostName.Should().BeFalse(); // The default is "false".
+                        }
+                        else
+                        {
+                            // Other authenticators do not contain mechanism properties
+                        }
+                    }
+                    else
+                    {
+                        foreach (var expectedMechanismProperty in expectedMechanismProperties.AsBsonDocument)
+                        {
+                            var mechanismName = expectedMechanismProperty.Name;
+                            switch (mechanismName)
+                            {
+                                case "SERVICE_NAME":
+                                    var serviceName = authenticator._mechanism()._serviceName();
+                                    serviceName.Should().Be(ValueToString(expectedMechanismProperty.Value));
+                                    break;
+                                case "CANONICALIZE_HOST_NAME":
+                                    var canonicalizeHostName = authenticator._mechanism()._canonicalizeHostName();
+                                    canonicalizeHostName.Should().Be(expectedMechanismProperty.Value.ToBoolean());
+                                    break;
+                                default:
+                                    throw new Exception($"Invalid mechanism property '{mechanismName}'.");
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private void AssertInvalid(Exception ex, BsonDocument definition)
+        {
+            if (definition["valid"].ToBoolean())
+            {
+                throw new AssertionException($"The connection string '{definition["uri"]}' should be valid.", ex);
+            }
+        }
+
+        private string ValueToString(BsonValue value)
+        {
+            return value == BsonNull.Value ? null : value.ToString();
+        }
+
+        // nested types
+        private class TestCaseFactory : JsonDrivenTestCaseFactory
+        {
+            protected override string PathPrefix => "MongoDB.Driver.Tests.Specifications.auth.tests.";
+        }
+    }
+
+    internal static class AuthenticatorReflector
+    {
+        public static object _mechanism(this IAuthenticator obj) => Reflector.GetFieldValue(obj, nameof(_mechanism));
```

Comment:
I see two problems with this class:

1. It doesn't really make sense to reflect over an interface (we should reflect over types, not interfaces)
2. The return type is object (let's return the actual type)

Reply:
The type returned by `ToAuthenticator()` method is `IAuthenticator`. There are several classes which implement this interface: `DefaultAuthenticator` `PlainAuthenticator`, `MongoDBCRAuthenticator`, `ScramSha1Authenticator`, `ScramSha256Authenticator` and `GssapiAuthenticator`. Only `GssapiAuthenticator` has `_mechanism` property, but this can potentially change in future, so I'm aiming for the common implementation.
See the response regarding `object` return type below in `GssapiMechanismReflector`.

Reply:
I don't think a reflector should allow you to reflect on arbitrary types returning arbitrary values.

A reflector should be over a specific type and return strongly typed values.

Since the _mechanism type is private and we can't use it here, let's treat the private nested type as if it's part of the parent type and handle both reflection steps without exposing the un-typed intermediate object:

    internal static class GssapiAuthenticatorReflector
    {
        public static bool _mechanism_canonicalizeHostName(this GssapiAuthenticator obj)
        {
            var mechanism = _mechanism(obj);
            var canonicalizeHostNameInfo = mechanism.GetType().GetField("_canonicalizeHostName", BindingFlags.NonPublic | BindingFlags.Instance);
            return (bool)canonicalizeHostNameInfo.GetValue(mechanism);
        }

        public static string _mechanism_serviceName(this GssapiAuthenticator obj)
        {
            var mechanism = _mechanism(obj);
            var serviceNameInfo = mechanism.GetType().GetField("_serviceName", BindingFlags.NonPublic | BindingFlags.Instance);
            return (string)serviceNameInfo.GetValue(mechanism);
        }

        private static object _mechanism(GssapiAuthenticator obj)
        {
            var mechanismInfo = typeof(SaslAuthenticator).GetField("_mechanism", BindingFlags.NonPublic | BindingFlags.Instance);
            return mechanismInfo.GetValue(obj);
        }
    }